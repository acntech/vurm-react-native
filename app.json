{
  "expo": {
    "name": "Vurm",
    "slug": "Vurm",
    "version": "1.0.0",
    "orientation": "portrait",
    "icon": "./assets/icon.png",
    "splash": {
      "image": "./assets/vurmLogo.png",
      "resizeMode": "contain",
      "backgroundColor": "#ffffff"
    },
    "updates": {
      "fallbackToCacheTimeout": 0
    },
    "assetBundlePatterns": [
      "**/*"
    ],
    "ios": {
      "supportsTablet": true,
      "bundleIdentifier": "com.chrisbso.Vurm"
    },
    "android": {
      "adaptiveIcon": {
        "foregroundImage": "./assets/adaptive-icon.png",
        "backgroundColor": "#FFFFFF"
      },
      "package": "com.chrisbso.Vurm"
    },
    "web": {
      "favicon": "./assets/favicon.png"
    },
    "facebookScheme": "fb498173242097723",
    "facebookAppId": "498173242097723",
    "facebookDisplayName": "Vurm",
    "facebookAdvertiserIDCollectionEnabled": false,
    "facebookAutoLogAppEventsEnabled": false
  }
}
