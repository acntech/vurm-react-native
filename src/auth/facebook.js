// Import the functions you need from the SDKs you need
import { auth } from "./firebase";
import { FacebookAuthProvider, signInWithCredential } from "firebase/auth";
import * as Facebook from "expo-facebook";

async function loginWithFacebook(setUser) {
  const appId = "498173242097723";
  await Facebook.initializeAsync({ appId });

  const { type, token } = await Facebook.logInWithReadPermissionsAsync({
    permissions: ["public_profile"],
  });

  if (type === "success") {
    // Build Firebase credential with the Facebook access token.
    const credential = FacebookAuthProvider.credential(token);

    // Sign in with credential from the Facebook user.
    signInWithCredential(auth, credential)
      .then(() => setUser(auth.currentUser))
      .catch((error) => {
        console.log(error);
      });
  }
  // return auth;
}

export default loginWithFacebook;
