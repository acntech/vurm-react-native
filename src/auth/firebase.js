import { initializeApp } from "firebase/app";
import { getAuth, onAuthStateChanged } from "firebase/auth";

const firebaseConfig = {
  apiKey: "AIzaSyB4i87NzfL1Pk_V2UpD5FSe1vwWV3-LBNs",
  authDomain: "leaderboard-rn.firebaseapp.com",
  databaseURL: "https://leaderboard-rn-default-rtdb.firebaseio.com",
  projectId: "leaderboard-rn",
  storageBucket: "leaderboard-rn.appspot.com",
  messagingSenderId: "1014828410467",
  appId: "1:1014828410467:web:35d753c749c30ee2e77a56",
  measurementId: "G-8D8MHJ3FZT",
};
//Initialize Firebase Application with our config
initializeApp(firebaseConfig);

// Listen for authentication state to change.
onAuthStateChanged(getAuth(), (user) => {
  if (user != null) {
    console.log("We are authenticated now!");
  }
});

export const auth = getAuth();
