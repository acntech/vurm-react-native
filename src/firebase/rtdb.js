import { getDatabase, ref, set, get, remove, child } from "firebase/database";

const getReference = (path) => {
  try {
    const db = getDatabase();
    return ref(db, path);
  } catch (error) {
    console.log(error);
  }
};

export const getUsersReference = () => {
  return getReference("users/");
};

export const getUserReference = (user) => {
  return getReference("users/" + user.uid);
};

export const getUserData = (user) => {
  return get(getUserReference(user));
};

export const deleteUserData = (user) => {
  if (user != null) {
    remove(getUserReference(user));
  }
};

export const setUserData = async (user, data) => {
  if (user != null) {
    const userReference = getUserReference(user);
    await set(userReference, data, (error) => {
      if (error) {
        console.log(error);
      }
    });
  }
};

export const getUserProperty = async (
  user,
  propertyName,
  defaultValue,
  setCallback
) => {
  let property = null;
  if (user != null) {
    const db = getDatabase();
    const reference = ref(db, "users/");
    await get(child(reference, user.uid))
      .then((snapshot) => {
        if (snapshot.exists()) {
          property = snapshot.val()[propertyName];
        } else {
          property = defaultValue;
        }
      })
      .catch((error) => {
        console.log(error);
      });
  }
  setCallback(property);
  return property;
};

export const extractDataFromUsersSnapshot = (usersSnapshot, userUid, topN) => {
  console.log("extract");
  let processedData = [];
  if (!usersSnapshot.hasChildren()) {
    return processedData;
  }
  let negRank = 0;
  let idx = 0;
  let prevScore = 0;
  let adjustedTopN = Object.keys(usersSnapshot.val()).length - topN - 1;

  usersSnapshot.forEach((child) => {
    const childValue = child.val();
    if (prevScore < childValue?.score) {
      negRank--;
      prevScore = childValue.score;
    }
    const youIndicator = child?.key == userUid ? " (You)" : "";
    childData = [
      negRank,
      childValue?.name + youIndicator,
      childValue?.score,
      child?.key,
    ];
    if (idx > adjustedTopN || child?.key == userUid) {
      processedData.push(childData);
    }
    idx++;
  });

  if (processedData.length > 0) {
    negRankIdx = 0;
    minNegRank = processedData[processedData.length - 1][negRankIdx];
    for (let i = 0; i < processedData.length; i++) {
      processedData[i][negRankIdx] += Math.abs(minNegRank) + 1;
    }
  }
  return processedData.reverse();
};
