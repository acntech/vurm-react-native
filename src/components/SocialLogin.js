import { View, Text, Button, Alert, StyleSheet } from "react-native";
import { signOut } from "firebase/auth";
import { onValue, orderByChild, query } from "firebase/database";
import { auth } from "../firebase/auth";
import { useState, useEffect } from "react";
import { deleteUserData, getUsersReference } from "../firebase/rtdb";
import { extractLeaderboardDataFromUsersSnapshot } from "./Leaderboard";
import HorizontalSeparator from "./HorizontalSeparator";
import FacebookLoginButton from "./FacebookLoginButton";
import GoogleLoginButton from "./GoogleLoginButton";

export default SocialLogin = () => {
  const [user, setUser] = useState(auth.currentUser);
  const [loginInProgress, setLoginInProgress] = useState(false);
  const [deleteMyDataButtonDisabled, setDeleteMyDataButtonDisabled] =
    useState(true);

  useEffect(() => {
    const usersReferenceOrderedByScore = query(
      getUsersReference(),
      orderByChild("score")
    );

    if (user) {
      const unsubscribe = onValue(
        usersReferenceOrderedByScore,
        (usersSnapshot) => {
          const userLeaderboardData = extractLeaderboardDataFromUsersSnapshot(
            usersSnapshot,
            user.uid,
            0
          );
          !userLeaderboardData.length
            ? setDeleteMyDataButtonDisabled(true)
            : setDeleteMyDataButtonDisabled(false);
        }
      );
      return unsubscribe;
    }
  }, []);

  const onPressSignOut = () => {
    signOut(auth)
      .then(() => {
        setUser(auth.currentUser);
      })
      .catch((error) => {
        Alert.alert("Failed to sign out", [
          {
            text: "OK",
          },
        ]);
      });
  };
  const onPressDeleteMyData = () => {
    Alert.alert(
      "Are you sure you wish to delete your data?",
      "This can't be undone",
      [
        {
          text: "Cancel",
          style: "cancel",
        },
        {
          text: "Delete",
          onPress: () => {
            deleteUserData(auth.currentUser);
            setDeleteMyDataButtonDisabled(true);
          },
          style: "destructive",
        },
      ]
    );
  };

  return (
    <View>
      {!user ? (
        <View>
          <FacebookLoginButton
            setUser={setUser}
            setDisabled={setLoginInProgress}
            buttonViewStyle={{ alignSelf: "center" }}
            disabled={loginInProgress}
          />
          <GoogleLoginButton
            setUser={setUser}
            setDisabled={setLoginInProgress}
            buttonViewStyle={{ alignSelf: "center" }}
            disabled={loginInProgress}
          ></GoogleLoginButton>
        </View>
      ) : (
        <View>
          {/* <HorizontalSeparator
            text={` ${user.displayName}  `}
          ></HorizontalSeparator> */}
          {/* 
          <Text style={{ alignSelf: "center" }}>
            Signed in as {user?.displayName}
          </Text> */}

          <Button
            style={{ alignSelf: "center" }}
            title={"Sign Out"}
            onPress={onPressSignOut}
          ></Button>
          {!deleteMyDataButtonDisabled ? (
            <Button
              style={{ alignSelf: "center" }}
              title={"Delete My Data"}
              color={"red"}
              onPress={onPressDeleteMyData}
            ></Button>
          ) : (
            <Text style={{ alignSelf: "center" }}>
              No data stored for {user.displayName}
            </Text>
          )}
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  // head: { height: 40, backgroundColor: "coral" },
  text: { margin: 6, color: "coral", textAlign: "center" },
  wrapper: { flexDirection: "row" },
});
