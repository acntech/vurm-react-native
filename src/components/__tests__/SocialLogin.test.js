import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import SocialLogin from "../SocialLogin";
import { auth } from "../../firebase/auth";
import { initializeApp, deleteApp, getApps } from "firebase/app";
import * as database from "firebase/database";
import { onValue, orderByChild, query } from "firebase/database";
jest.useFakeTimers();
jest.mock("react-native/Libraries/Animated/NativeAnimatedHelper");

// const mockPromptAsync = jest.fn(() => Promise.resolve());
// auth = { currentUser: { uid: "testUid", displayName: "Test Person" } };

// jest.spyOn(auth, "currentUser").mockImplementation(() => ({
//   uid: "testUid",
//   displayName: "Test Person",
// }));
jest.mock("../../firebase/auth", () => ({
  auth: null,
}));
jest.mock("../FacebookLoginButton");
jest.mock("../GoogleLoginButton");

jest.mock("firebase/auth", () => ({
  ...jest.requireActual("firebase/auth"),
  onAuthStateChanged: jest.fn(() => () => {}),
}));

jest.mock("firebase/database");
database.onValue = jest.fn(() => () => {});
database.orderByChild = jest.fn();
database.query = jest.fn();

describe("<SocialLogin />", () => {
  describe("renders correctly", () => {
    it("before login", () => {
      // Arrange & Act
      const tree = render(<SocialLogin />).toJSON();

      // Assert
      expect(tree).toMatchSnapshot();
    });
  });

  it("after login without user data", async () => {
    const tree = render(<SocialLogin />);
    const { getByText } = tree;
    const toClick = getByText("Sign in with Google");

    // Act
    fireEvent(toClick, "press");

    await waitFor(() => expect(getByText("Sign Out")).toBeTruthy());
  });

  //   it("after login with user data", () => {
  //     expect(true).toBe(true);
  //   });
});
