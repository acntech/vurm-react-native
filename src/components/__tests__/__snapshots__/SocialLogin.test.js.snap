// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SocialLogin /> renders correctly before login 1`] = `
<View>
  <View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#485a96",
          "borderColor": "#fff",
          "borderRadius": 5,
          "borderWidth": 0.5,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "center",
          "margin": 5,
          "marginBottom": 0,
          "opacity": 1,
          "position": "relative",
          "width": 220,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "alignSelf": "center",
            "height": 25,
            "marginLeft": 15,
            "padding": 10,
            "resizeMode": "stretch",
            "width": 25,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#fff",
            "marginLeft": 20,
            "marginRight": 20,
          }
        }
      >
        Sign in with Facebook
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#fff",
          "borderRadius": 5,
          "borderWidth": 0.5,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "center",
          "margin": 5,
          "opacity": 1,
          "width": 220,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 25,
            "marginLeft": 15,
            "padding": 10,
            "resizeMode": "stretch",
            "width": 25,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#575757",
            "marginLeft": 15,
            "marginRight": 20,
          }
        }
      >
        Sign in with Google
      </Text>
    </View>
  </View>
</View>
`;
